pkg/controllers/customcluster_controller.go

```diff
@@ -228,719 +240,249 @@ func (r *CustomClusterController) Reconcile(ctx context.Context, req ctrl.Reques
+		log.Info("phase changes", "prevPhase", customCluster.Status.Phase, "currentPhase", v1alpha1.ProvisionFailedPhase)
+		customCluster.Status.Phase = v1alpha1.ProvisionFailedPhase
+		conditions.MarkFalse(customCluster, v1alpha1.ReadyCondition, v1alpha1.InitWorkerRunFailedReason,
+			clusterv1.ConditionSeverityWarning, "init worker run failed %s/%s", customCluster.Namespace, customCluster.Name)
 
-	if workerPod.Status.Phase == corev1.PodFailed {
-		customCluster.Status.Phase = v1alpha1.UnknownPhase
-
-		conditions.MarkFalse(customCluster, v1alpha1.ScaledDownCondition, v1alpha1.ScaleDownWorkerRunFailedReason,
-			clusterv1.ConditionSeverityWarning, "scale down worker run failed %s/%s", customCluster.Namespace, customCluster.Name)
-
-		log.Info(fmt.Sprintf("customCluster's phase changes from %s to %s", string(v1alpha1.UnknownPhase), string(v1alpha1.UnknownPhase)))
 		return ctrl.Result{}, nil
 	}
 
 	return ctrl.Result{}, nil
 }
 
-// reconcileHandleDeleting determine whether customCluster go to reconcileDeleteResource or enter DeletingFailed phase when current phase is Deleting.
-func (r *CustomClusterController) reconcileHandleDeleting(ctx context.Context, customCluster *v1alpha1.CustomCluster, customMachine *v1alpha1.CustomMachine) (ctrl.Result, error) {
+// reconcileDelete handle cluster deletion.
+func (r *CustomClusterController) reconcileDelete(ctx context.Context, customCluster *v1alpha1.CustomCluster, customMachine *v1alpha1.CustomMachine, kcp *controlplanev1.KubeadmControlPlane) (ctrl.Result, error) {
 	log := ctrl.LoggerFrom(ctx)
 
-	terminateWorker := &corev1.Pod{}
-	terminateWorkerKey := generateWorkerKey(customCluster, CustomClusterTerminateAction)
+	// Delete the manager worker pods first if there are still any running.
+	if err := r.deleteWorkerPods(ctx, customCluster); err != nil {
+		log.Error(err, "failed to delete worker pods", "name", customCluster.Name, "namespace", customCluster.Namespace)
+		return ctrl.Result{}, err
+	}
 
-	if err := r.Client.Get(ctx, terminateWorkerKey, terminateWorker); err != nil {
-		if apierrors.IsNotFound(err) {
-			log.Info("terminate worker does not exist, turn to reconcileVMsTerminate to create a new one", "worker", terminateWorkerKey)
-			return r.reconcileVMsTerminate(ctx, customCluster)
-		}
-		log.Error(err, "failed to get terminate worker. maybe it has been deleted", "worker", terminateWorkerKey)
-		return ctrl.Result{RequeueAfter: RequeueAfter}, err
+	// Create the termination worker to handle the cluster deletion.
+	terminateWorker, err1 := r.ensureWorkerPodCreated(ctx, customCluster, CustomClusterTerminateAction, KubesprayTerminateCMD, generateClusterHostsName(customCluster), generateClusterConfigName(customCluster))
+	if err1 != nil {
+		conditions.MarkFalse(customCluster, v1alpha1.TerminatedCondition, v1alpha1.FailedCreateTerminateWorker,
+			clusterv1.ConditionSeverityWarning, "terminate worker not ready %s/%s.", customCluster.Namespace, customCluster.Name)
```

Comment:
why `worker not ready`, isn't it failed to create ？

Reply:
ok， it need rename